src/chai/SharedPtrManager.hpp

```diff
@@ -0,0 +1,442 @@
+   * \return Pointer to data in the current execution space.
+   */
+  CHAISHAREDDLL_API void* move(void* pointer,
+                               msp_pointer_record* pointer_record,
+                               ExecutionSpace = NONE, bool = false);
+
+  /*!
+   * \brief Register a touch of the pointer in the current execution space.
+   *
+   * \param pointer Raw pointer to register a touch of.
+   */
+  CHAISHAREDDLL_API void registerTouch(msp_pointer_record* pointer_record);
+
+  /*!
+   * \brief Register a touch of the pointer in the given execution space.
+   *
+   * The pointer doesn't need to exist in the space being touched.
+   *
+   * \param pointer Raw pointer to register a touch of.
+   * \param space Space to register touch.
+   */
+  CHAISHAREDDLL_API void registerTouch(msp_pointer_record* pointer_record, ExecutionSpace space);
+
+  /*!
+   * \brief Make a new allocation of the data described by the msp_pointer_record in
+   * the given space.
+   *
+   * \param pointer_record
+   * \param space Space in which to make the allocation.
+   */
+  CHAISHAREDDLL_API void allocate(msp_pointer_record* pointer_record, ExecutionSpace space = CPU);
+
+  /*!
+   * \brief Reallocate data.
+   *
+   * Data is reallocated in all spaces this pointer is associated with.
+   *
+   * \param ptr Pointer to address to reallocate
+   * \param elems The number of elements to allocate.
+   * \tparam T The type of data to allocate.
+   *
+   * \return Pointer to the allocated memory.
+   */
+//  template <typename T>
+//  void* reallocate(void* pointer,
+//                   size_t elems,
+//                   msp_pointer_record* record);
+
+  /*!
+   * \brief Set the default space for new ManagedArray allocations.
+   *
+   * ManagedArrays allocated without an explicit ExecutionSpace argument will
+   * be allocated in space after this routine is called.
+   *
+   * \param space New space for default allocations.
+   */
+  CHAISHAREDDLL_API void setDefaultAllocationSpace(ExecutionSpace space);
+
+  /*!
+   * \brief Get the currently set default allocation space.
+   *
+   * See also setDefaultAllocationSpace.
+   *
+   * \return Current default space for allocations.
+   */
+  CHAISHAREDDLL_API ExecutionSpace getDefaultAllocationSpace();
+
+  /*!
+   * \brief Free allocation(s) associated with the given msp_pointer_record.
+   *        Default (space == NONE) will free all allocations and delete
+   *        the pointer record.
+   */
+  CHAISHAREDDLL_API void free(msp_pointer_record* pointer, ExecutionSpace space = NONE);
+
+#if defined(CHAI_ENABLE_PICK)
+  template <typename T>
+   T_non_const<T> pick(T* src_ptr, size_t index);
+
+  template <typename T>
+   void set(T* dst_ptr, size_t index, const T& val);
+#endif
+
+  /*!
+   * \brief Get the size of the given pointer.
+   *
+   * \param pointer Pointer to find the size of.
+   * \return Size of pointer.
+   */
+  //CHAISHAREDDLL_API size_t getSize(void* pointer);
```

Comment:
I would prefer that unimplemented functions or ones we aren't sure are applicable be removed and we just add an issue to track them.

Reply:
I will go through and review / remove all commented code added in this PR